src/btree/bt_read.c

```diff
@@ -255,89 +249,65 @@ __instantiate_lookaside(WT_SESSION_IMPL *session, WT_REF *ref)
-     * maximum timestamp and the transaction id as part of the key search. This search can return
-     * either the last record of the search key or the first record of the next key.
+     * An important point to note is that the keys for a given page are NOT necessarily next to each
+     * other in the lookaside table since we can specify our own ordering for a given table with a
+     * custom collator. Therefore, we need to make use of the keys that we have stored in-memory
+     * last time we evicted to instantiate each key.
      *
-     * Once we instantiate the most recent record for a key, the search continues to the next key.
-     * Since we're pointing at the max timestamp and transaction id pairing for a given key, the
-     * next entry is guaranteed to be for the next key in the LAS and again set the search for the
-     * next key with maximum timestamp and transaction id.
+     * During instantiation, we iterate over our set of keys from eviction. If the key memento has a
+     * specific txn id that isn't "aborted" then it indicates that birthmark update should be
+     * instantiated for that key. Otherwise it is just an indicator that we need to search the
+     * lookaside for that particular key.
      */
     cache->las_reader = true;
     __wt_readlock(session, &cache->las_sweepwalk_lock);
     cache->las_reader = false;
     notused = size = total_incr = 0;
-    first_scan = true;
     locked = true;
-    las_cursor->set_key(las_cursor, las_btree_id, &page_las->min_las_key, WT_TS_MAX, WT_TXN_MAX);
-    for (; ret == 0; ret = las_cursor->next(las_cursor)) {
-        if (!first_scan) {
-            WT_ERR(las_cursor->get_key(
-              las_cursor, &las_btree_id, &next_las_key, &las_timestamp, &las_txnid));
-
-            /* Set the key to check with maximum timestamp and transaction id. */
-            las_cursor->set_key(las_cursor, las_btree_id, &next_las_key, WT_TS_MAX, WT_TXN_MAX);
-        }
-        first_scan = false;
-
-        WT_ERR(las_cursor->search_near(las_cursor, &exact));
-        /*
-         * Check whether the search_near returned the same key. If it returns the next key, continue
-         * with the previous record.
-         */
-        if (exact > 0)
-            WT_ERR(las_cursor->prev(las_cursor));
-
-        WT_ERR(
-          las_cursor->get_key(las_cursor, &las_btree_id, &las_key, &las_timestamp, &las_txnid));
-
-        /* Stop before crossing over to the next btree. */
-        if (las_btree_id != S2BT(session)->id)
-            break;
-
-        WT_ERR(__wt_compare(session, NULL, &las_key, &page_las->min_las_key, &cmp));
-        /* We got a key that is less than the minimum key of the page. */
-        if (cmp < 0)
-            break;
+    for (i = 0; i < page_las->birthmarks_cnt; ++i) {
+        if (page_las->birthmarks[i].txnid != WT_TXN_ABORTED) {
+            WT_ERR(__create_birthmark_upd(session, &page_las->birthmarks[i], &size, &upd));
```

Comment:
please add a comment saying this is a birthmark

Reply:
535a348